lib/checksum.c

```diff
@@ -42,794 +41,858 @@
-    case RM_DIGEST_SHA1:
-        self = g_slice_new0(RmDigest);
-        self->bytes = digest->bytes;
-        self->type = digest->type;
-        self->glib_checksum = g_checksum_copy(digest->glib_checksum);
-        break;
-    case RM_DIGEST_SHA3_256:
-    case RM_DIGEST_SHA3_384:
-    case RM_DIGEST_SHA3_512:
-        self = g_slice_new0(RmDigest);
-        self->bytes = digest->bytes;
-        self->type = digest->type;
-        self->sha3_ctx = g_slice_alloc0(sizeof(sha3_context));
-        memcpy(self->sha3_ctx, digest->sha3_ctx, sizeof(sha3_context));
-        break;
-    case RM_DIGEST_BLAKE2S:
-        BLAKE_COPY(blake2s);
-        break;
-    case RM_DIGEST_BLAKE2B:
-        BLAKE_COPY(blake2b);
-        break;
-    case RM_DIGEST_BLAKE2SP:
-        BLAKE_COPY(blake2sp);
-        break;
-    case RM_DIGEST_BLAKE2BP:
-        BLAKE_COPY(blake2bp);
-        break;
-    case RM_DIGEST_SPOOKY:
-    case RM_DIGEST_SPOOKY32:
-    case RM_DIGEST_SPOOKY64:
-    case RM_DIGEST_MURMUR:
-    case RM_DIGEST_CITY:
-    case RM_DIGEST_CITY256:
-    case RM_DIGEST_MURMUR256:
-    case RM_DIGEST_CITY512:
-    case RM_DIGEST_MURMUR512:
-    case RM_DIGEST_XXHASH:
-    case RM_DIGEST_FARMHASH:
-    case RM_DIGEST_BASTARD:
-    case RM_DIGEST_CUMULATIVE:
-    case RM_DIGEST_EXT:
-        self = rm_digest_new(digest->type, 0, 0, digest->bytes, FALSE);
-
-        if(self->checksum && digest->checksum) {
-            memcpy((char *)self->checksum, (char *)digest->checksum, self->bytes);
-        }
+static gpointer rm_init_digest_type_table(GHashTable **code_table) {
+    *code_table = g_hash_table_new(g_str_hash, g_str_equal);
+    for(RmDigestType type = 1; type < RM_DIGEST_SENTINEL; type++) {
+        rm_digest_table_insert(*code_table, (char *)rm_digest_get_interface(type)->name,
+                               type);
+    }
+
+    /* add some synonyms */
+    rm_digest_table_insert(*code_table, "sha3", RM_DIGEST_SHA3_256);
+    rm_digest_table_insert(*code_table, "highway", RM_DIGEST_HIGHWAY256);
+
+    return NULL;
+}
 
-        break;
-    case RM_DIGEST_PARANOID:
-    default:
-        rm_assert_gentle_not_reached();
+///////////////////////////////////////
+//           RMDIGEST API            //
+///////////////////////////////////////
+
+RmDigestType rm_string_to_digest_type(const char *string) {
+    static GHashTable *code_table = NULL;
+    static GOnce table_once = G_ONCE_INIT;
+
+    if(string == NULL) {
+        return RM_DIGEST_UNKNOWN;
     }
 
-    return self;
+    g_once(&table_once, (GThreadFunc)rm_init_digest_type_table, &code_table);
+
+    gchar *lower_key = g_utf8_strdown(string, -1);
+    RmDigestType code = GPOINTER_TO_UINT(g_hash_table_lookup(code_table, lower_key));
+    g_free(lower_key);
+
+    return code;
+}
+
+const char *rm_digest_type_to_string(RmDigestType type) {
+    const RmDigestInterface *interface = rm_digest_get_interface(type);
+    return interface->name;
+}
+
+/*  TODO: update or remove? */
```

Comment:
Remove. Was a nice thought, but since ``rmlint --hash`` will probably never be a widely used tool, it does not make sense. This was the idea behind it (self describing hashes): https://github.com/multiformats/multihash

Reply:
Done.